tools/lemmacmd.go

```diff
@@ -0,0 +1,267 @@
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"golang.org/x/crypto/pbkdf2"
+
+	"github.com/mailgun/lemma/random"
+	"github.com/mailgun/lemma/secret"
+)
+
+var (
+	inputpath  = flag.String("in", "", "path to file to be read in")
+	outputpath = flag.String("out", "", "path to file to be written out")
+	mode       = flag.String("mode", "", "mode: enc[rypt] or dec[rypt]")
+	keypath    = flag.String("keypath", "", "path to base64-encoded 32-byte key on disk, if no path is provided, a passphrase will be used")
+	itercount  = flag.Int("itercount", 524288, "if a passphrase is used, iteration count for pbkdf#2")
+)
+
+type EncodedCiphertext struct {
+	KeySalt         string `json:"key_salt,omitempty"`
+	KeyIter         int    `json:"key_iter_count,omitempty"`
+	CiphertextNonce string `json:"ciphertext_nonce"`
+	Ciphertext      string `json:"ciphertext"`
+}
+
+func main() {
+	flag.Parse()
+
+	// validate required parameters
+	if flag.NFlag() == 0 {
+		fmt.Printf("usage: lemmacmd [flags]\n")
+		fmt.Printf("Flags:\n")
+		flag.PrintDefaults()
+		os.Exit(255)
+	}
+	if *inputpath == "" {
+		fmt.Printf("lemmacmd: input path required\n")
+		os.Exit(255)
+	}
+	if *outputpath == "" {
+		fmt.Printf("lemmacmd: output path required\n")
+		os.Exit(255)
+	}
+
+	switch *mode {
+	case "enc":
+		keySalt, key, err := generateEncryptionKey(*keypath, *itercount)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to generate or read in key: %v\n", err)
+			os.Exit(255)
+		}
+
+		secretService, err := secret.New(&secret.Config{KeyBytes: key})
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to build lemma.secret: %v\n", err)
+			os.Exit(255)
+		}
+
+		plaintextBytes, err := ioutil.ReadFile(*inputpath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to read plaintext file %q: %v\n", *inputpath, err)
+			os.Exit(255)
+		}
+
+		sealedData, err := secretService.Seal(plaintextBytes)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to seal plaintext: %v\n", err)
+			os.Exit(255)
+		}
+
+		err = writeCiphertext(keySalt, *itercount, sealedData, *outputpath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to write sealed data to disk: %v\n", err)
+			os.Exit(255)
+		}
```

Comment:
What do you think of pulling the code in these cases out into their own functions? e.g.

```go
switch *mode {
case "enc":
        encrypt()
case "dec":
        decrypt()
default:
        fmt.Printf("lemmacmd: invalid mode provided: %v\n", *mode)
        os.Exit(255)
}
````